pgdedupe/mssql/cli.py

```diff
@@ -0,0 +1,526 @@
+    config['columns'] = ', '.join(columns)
+    config['all_columns'] = ', '.join(columns | set(['_unique_id']))
+    return config
+
+
+def unicode_to_str(data):
+    if data == "":
+        data = None
+    if PYTHON_VERSION < 3:
+        if isinstance(data, basestring):
+            return str(data)
+        elif isinstance(data, collections.Mapping):
+            return dict(map(unicode_to_str, data.iteritems()))
+        elif isinstance(data, collections.Iterable):
+            return type(data)(map(unicode_to_str, data))
+        else:
+            return data
+    else:
+        if isinstance(data, str):
+            return str(data)
+        elif isinstance(data, collections.Mapping):
+            return dict(map(unicode_to_str, data.items()))
+        elif isinstance(data, collections.Iterable):
+            return type(data)(map(unicode_to_str, data))
+        else:
+            return data
+
+
+def preprocess(con, config):
+    c = con.cursor()
+
+    # Ensure the database has the schema and required functions
+    c.execute("""IF NOT EXISTS (SELECT * FROM sys.schemas WHERE name = '{schema}')
+                BEGIN
+                EXEC('CREATE SCHEMA {schema}')
+                END""".format(**config))
+    """
+    # Create an intarray-like idx function (https://wiki.postgresql.org/wiki/Array_Index):
+    c.execute(""CREATE OR REPLACE FUNCTION {schema}.idx(anyarray, anyelement)
+                   RETURNS INT A
+                 $$
+                   SELECT i FROM (
+                      SELECT generate_series(array_lower($1,1),array_upper($1,1))
+                   ) g(i)
+                   WHERE $1[i] = $2
+                   LIMIT 1;
+                 $$ LANGUAGE SQL IMMUTABLE;".format(**config))
+    """
+    # Do an initial first pass and merge all exact duplicates
+    c.execute("IF OBJECT_ID('{schema}.entries_unique', 'U') IS NOT NULL "
+              "DROP TABLE {schema}.entries_unique".format(**config))
+    c.execute("""SELECT DISTINCT {case_sensitive_columns} INTO {schema}.entries_unique
+                    FROM {table}
+                    WHERE ({filter_condition})""".format(**config))
+
+    c.execute("ALTER TABLE {schema}.entries_unique "
+              " ADD _unique_id INT IDENTITY(1,1) PRIMARY KEY".format(**config))
+
+    c.execute("IF OBJECT_ID('{schema}.entries_src_ids', 'U') IS NOT NULL "
+              "DROP TABLE {schema}.entries_src_ids".format(**config))
+
+    c.execute("CREATE INDEX table_column_idx "
+              " ON {table} ({key}, {columns})".format(**config))
+    c.execute("CREATE INDEX {schema}_entries_unique_column_idx "
+              " ON {schema}.entries_unique (_unique_id, {columns})".format(**config))
+    c.execute("""SELECT t._unique_id, {table}.{key} INTO {schema}.entries_src_ids
+                    FROM {schema}.entries_unique as t, {table}
+                    WHERE
+                    BINARY_CHECKSUM({column_select_table}) = BINARY_CHECKSUM({column_select_t})"""
```

Comment:
Is this a valid optimization?  I'm a bit worried about collisions here over large tables.